br/pkg/lightning/restore/get_pre_info.go

```diff
@@ -0,0 +1,801 @@
+func (p *PreRestoreInfoGetterImpl) GetAllTableStructures(ctx context.Context) (map[string]*checkpoints.TidbDBInfo, error) {
+	var (
+		dbInfos map[string]*checkpoints.TidbDBInfo
+		err     error
+	)
+	dbInfosVal := ctx.Value(preInfoGetterKeyTableStructsCache)
+	if dbInfosVal != nil {
+		if v, ok := dbInfosVal.(map[string]*checkpoints.TidbDBInfo); ok {
+			dbInfos = v
+		}
+	}
+	if dbInfos != nil {
+		return dbInfos, nil
+	}
+	dbInfos, err = LoadSchemaInfo(ctx, p.dbMetas, func(ctx context.Context, dbName string) ([]*model.TableInfo, error) {
+		return p.getTableStructuresByFileMeta(ctx, p.mdDBMetaMap[dbName])
+	})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return dbInfos, nil
+}
+
+func (p *PreRestoreInfoGetterImpl) getTableStructuresByFileMeta(ctx context.Context, dbSrcFileMeta *mydump.MDDatabaseMeta) ([]*model.TableInfo, error) {
+	dbName := dbSrcFileMeta.Name
+	currentTableInfosFromDB, err := p.targetInfoGetter.FetchRemoteTableModels(ctx, dbName)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	currentTableInfosMap := make(map[string]*model.TableInfo)
+	for _, tblInfo := range currentTableInfosFromDB {
+		currentTableInfosMap[tblInfo.Name.L] = tblInfo
+	}
+	resultInfos := make([]*model.TableInfo, len(dbSrcFileMeta.Tables))
+	for i, tableFileMeta := range dbSrcFileMeta.Tables {
+		if curTblInfo, ok := currentTableInfosMap[strings.ToLower(tableFileMeta.Name)]; ok {
+			resultInfos[i] = curTblInfo
+			continue
+		}
+		createTblSQL, err := tableFileMeta.GetSchema(ctx, p.srcStorage)
+		if err != nil {
+			return nil, errors.Annotatef(err, "get create table statement from schema file error: %s", tableFileMeta.Name)
+		}
+		theTableInfo, err := newTableInfo(createTblSQL, 0)
+		if err != nil {
+			errMsg := "generate table info from SQL error"
+			log.L().Error(errMsg, zap.Error(err), zap.String("sql", createTblSQL), zap.String("table_name", tableFileMeta.Name))
+			return nil, errors.Annotatef(err, "%s: %s", errMsg, tableFileMeta.Name)
+		}
+		resultInfos[i] = theTableInfo
+	}
+	return resultInfos, nil
+}
+
+func newTableInfo(createTblSQL string, tableID int64) (*model.TableInfo, error) {
+	parser := parser.New()
+	astNode, err := parser.ParseOneStmt(createTblSQL, "", "")
+	if err != nil {
+		errMsg := "parse sql statement error"
+		log.L().Error(errMsg, zap.Error(err), zap.String("sql", createTblSQL))
+		return nil, errors.Trace(err)
+	}
+	sctx := mock.NewContext()
+	createTableStmt, ok := astNode.(*ast.CreateTableStmt)
+	if !ok {
+		return nil, errors.New("cannot transfer the parsed SQL as an CREATE TABLE statement")
+	}
+	info, err := ddl.MockTableInfo(sctx, createTableStmt, tableID)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	// set a auto_random bit if AUTO_RANDOM is set
+	setAutoRandomBits(info, createTableStmt.Cols)
+	info.State = model.StatePublic
+	return info, nil
+}
+
+func setAutoRandomBits(tblInfo *model.TableInfo, colDefs []*ast.ColumnDef) {
+	if tblInfo.PKIsHandle {
```

Comment:
maybe reverse this to reduce indent level, `colDef.Name.Name.L == pkColName.L && colDef.Tp.GetType() == mysql.TypeLonglong` too